add a new benchmark for volatiles release, migration and reabsoption

jdannberg opened the pull request:
This PR adds a new benchmark for water solubility. The new plugin in the benchmark folder can also serve as an example for how water solubility can be implemented in a plugin in ASPECT. 

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

naliboff commented on benchmarks/solubility/plugin/solubility.cc:
```diff
@@ -0,0 +1,409 @@
+                             Patterns::Double (0),
+                             "In case the operator splitting scheme is used, the porosity field can not "
+                             "be set to a new equilibrium melt fraction instantly, but the model has to "
+                             "provide a melting time scale instead. This time scale defines how fast melting "
```
```suggestion
                             "provide a fluid reaction time scale instead. This time scale defines how fast melting "
```

naliboff commented on benchmarks/solubility/plugin/solubility.cc:
```diff
@@ -0,0 +1,409 @@
+                             "be set to a new equilibrium melt fraction instantly, but the model has to "
+                             "provide a melting time scale instead. This time scale defines how fast melting "
+                             "happens, or more specifically, the parameter defines the time after which "
+                             "the deviation of the porosity from the equilibrium melt fraction will be "
```
```suggestion
                             "the deviation of the porosity from the equilibrium fluid fraction will be "
```

naliboff commented on benchmarks/solubility/plugin/solubility.cc:
```diff
@@ -0,0 +1,409 @@
+                             "provide a melting time scale instead. This time scale defines how fast melting "
+                             "happens, or more specifically, the parameter defines the time after which "
+                             "the deviation of the porosity from the equilibrium melt fraction will be "
+                             "reduced to a fraction of $1/e$. So if the melting time scale is small compared "
```
```suggestion
                             "reduced to a fraction of $1/e$. So if the fluid reaction time scale is small compared "
```

naliboff replied:
... comment applies to similar times below ...

naliboff commented on benchmarks/solubility/plugin/solubility.cc:
```diff
@@ -0,0 +1,409 @@
+    {
+      prm.enter_subsection("Material model");
+      {
+        prm.enter_subsection("Volatile model");
```
```suggestion
        prm.enter_subsection("Reactive fluid transport model");
```

naliboff commented on benchmarks/solubility/plugin/solubility.cc:
```diff
@@ -0,0 +1,409 @@
+    template <int dim>
+    void
+    Volatiles<dim>::
+    melt_fractions (const MaterialModel::MaterialModelInputs<dim> &in,
```
```suggestion
    fluid_fractions (const MaterialModel::MaterialModelInputs<dim> &in,
```

jdannberg replied:
Same here, this is not something I can change, this is defined by the interface.

naliboff commented on benchmarks/solubility/plugin/solubility.cc:
```diff
@@ -0,0 +1,409 @@
+{
+  namespace MaterialModel
+  {
+    ASPECT_REGISTER_MATERIAL_MODEL(Volatiles,
```
```suggestion
    ASPECT_REGISTER_MATERIAL_MODEL(Reactive Fluid Transport,
```

naliboff replied:
Another name could be `Reactive Fluids`

naliboff commented on benchmarks/solubility/plugin/solubility.cc:
```diff
@@ -0,0 +1,409 @@
+  namespace MaterialModel
+  {
+    ASPECT_REGISTER_MATERIAL_MODEL(Volatiles,
+                                   "volatiles",
```
```suggestion
                                   "reactive fluid transport",
```

naliboff commented on benchmarks/solubility/plugin/solubility.cc:
```diff
@@ -0,0 +1,409 @@
+  {
+    ASPECT_REGISTER_MATERIAL_MODEL(Volatiles,
+                                   "volatiles",
+                                   "Material model that can advect volatiles and contains "
```
```suggestion
                                   "A composited material model composited designed to both advect "
                                   "fluids and calculate equilibrium fluid fractions. It currently contains "
```

naliboff commented on benchmarks/solubility/plugin/solubility.cc:
```diff
@@ -0,0 +1,409 @@
+      // Fill the melt outputs if they exist.
+      MeltOutputs<dim> *melt_out = out.template get_additional_output<MeltOutputs<dim>>();
+
+      if (melt_out != nullptr)
```
Could this section of code be expanded to have a condition where the MeltOutputs are not filled, and instead the fluids are advected using the single-phase Darcy method? 

As an alternative, could a trick be done where the porosity given to the MeltOutputs is set to 0, such that is solves the normal equations for solid deformation and the fluids are advected with the darcy field?

jdannberg replied:
I think you could just switch off melt transport and then change the porosity to be of type darcy_field. I do not think this would require changing the code, but there may be some things I am not considering at the moment.

naliboff reviewed (commented):
@jdannberg - Thank you for this contribution, which will be really useful for a wide range of groups!

There are lots of comments, but nearly all related to parameter naming/descriptions and related documentation. In short, they all boil down to how generic to make the plugin at this stage, as well as in the future. Aside from these conventions, the code and example are in good shape, and can be merged with little modification.

Here are the general questions I have in mind:

1. Should this plugin be added in the material model section (rather than as a cookbook), or should that step occur later down the line? I think this should occur sooner rather than later for testing purposes, but it does not need to be now.

2. As the model can in theory be expanded to  deal with any type of geofluid, should the name be changed to something like `reactive fluid transport` or `reactive fluids`? My preference is for the former, as it indicates the material both deals with the reaction and advection of fluids.

3. Where possible, I think variable names containing melt or volatiles should be reduced to more generic names (fluids), unless they specifically are addressing a specific type of geofluid.

4. I left one comment/question about how the material model could be modified to use either the two-phase flow equations or the simpler darcy field method for advecting fluids.

To reiterate, I am fine if many of the suggested changes above occur when the code (or a copy of it) is moved to the material model section.

@danieldouglas92 and Ryan - what are your thoughts?

Thanks Juliane!

ryanstoner1 commented:
>     1. Should this plugin be added in the material model section (rather than as a cookbook), or should that step occur later down the line? I think this should occur sooner rather than later for testing purposes, but it does not need to be now.
> 

If adding this to the material model would help with numerical stability then the material model might be good to do sooner. If not then my vote is for whatever makes things the most modular.

>     2. As the model can in theory be expanded to  deal with any type of geofluid, should the name be changed to something like `reactive fluid transport` or `reactive fluids`? My preference is for the former, as it indicates the material both deals with the reaction and advection of fluids.
> 

A vote for `reactive fluid transport` for the same reasons. 

>     3. Where possible, I think variable names containing melt or volatiles should be reduced to more generic names (fluids), unless they specifically are addressing a specific type of geofluid.
> 

This would be more correct from a geochemistry perspective.

jdannberg commented on benchmarks/solubility/plugin/solubility.cc:
```diff
@@ -0,0 +1,409 @@
+    using namespace dealii;
+
+    /**
+     * Volatiles material model.
```
The way I envisioned this is that what I wrote here is really only the benchmark. If @danieldouglas92 wants to implement a more complex model, I thought he would just make a copy of this, because I was assuming that the two of you might want to change a lot of things and this is really just a minimal working example. 

You could then pick a more general name for your material model.

jdannberg commented on benchmarks/solubility/plugin/solubility.cc:
```diff
@@ -0,0 +1,409 @@
+         */
+        std::unique_ptr<MaterialModel::Interface<dim>> base_model;
+
+        double reference_rho_f;
```
I am adding a short description, but these are also all documented in the declare parameters function (since they are read from the input file).

jdannberg commented on benchmarks/solubility/plugin/solubility.cc:
```diff
@@ -0,0 +1,409 @@
+        double reference_permeability;
+        double alpha_phi;
+        double melt_compressibility;
+        double melting_time_scale;
```
Ha! When I saw your comment about making it more general, this is exactly the name I changed it to (before even seeing this comment)!

jdannberg commented on benchmarks/solubility/plugin/solubility.cc:
```diff
@@ -0,0 +1,409 @@
+         * @param melt_fractions Vector of doubles that is filled with the
+         * equilibrium melt fraction for each given input conditions.
+         */
+        virtual void melt_fractions (const MaterialModel::MaterialModelInputs<dim> &in,
```
This I cannot change because it is a function that is implemented in the `MeltFractionModel` (one of the classes that this class is derived from) with this name. But I am changing the documentation.

jdannberg commented on benchmarks/solubility/plugin/solubility.cc:
```diff
@@ -0,0 +1,409 @@
+        }
+
+      // Fill the melt outputs if they exist.
+      MeltOutputs<dim> *melt_out = out.template get_additional_output<MeltOutputs<dim>>();
```
The class or just the variable name?

jdannberg commented on benchmarks/solubility/plugin/solubility.cc:
```diff
@@ -0,0 +1,409 @@
+        }
+
+      ReactionRateOutputs<dim> *reaction_rate_out = out.template get_additional_output<ReactionRateOutputs<dim>>();
+      const unsigned int water_idx = this->introspection().compositional_index_for_name("water_content");
```
I thought of this benchmark as being something very specific, so I would prefer to keep it like that for this example, and @danieldouglas92 could change the name in his model.

jdannberg commented:
@naliboff Thank you for your review! I addressed your comments. For your more general questions:

> 1. Should this plugin be added in the material model section (rather than as a cookbook), or should that step occur later down the line? I think this should occur sooner rather than later for testing purposes, but it does not need to be now.

I would leave the current version as a benchmark, but put the new model @danieldouglas92 is making into the material model folder. This way, this simple model can serve as an example for how the implementation of such a model would work and illustrate the general concept, and the model Daniel is writing (that is actually applicable to Earth) can be used more easily by others. 

Also, tests can include files from the benchmarks folder (as I did in this PR) so I do not think it would make a difference for that. 

> 2. As the model can in theory be expanded to deal with any type of geofluid, should the name be changed to something like reactive fluid transport or reactive fluids? My preference is for the former, as it indicates the material both deals with the reaction and advection of fluids.

> 3. Where possible, I think variable names containing melt or volatiles should be reduced to more generic names (fluids), unless they specifically are addressing a specific type of geofluid.

Same here, my model is just for water, so everywhere where it said melt, I changed that, but where I referred to water and water content, I left it like that. When you make a new model that is more general, that model could get a different name.

naliboff reviewed (approved):
@jdannberg - Thanks for all of the changes, and agree the best path forward is to implement the other changes discussed (move to plugin to material model section, adjust name of material model) in the upcoming PR from @danieldouglas92.

naliboff commented:
@jdannberg - One of the the testers had an error while attempting to run the solubility cookbook test (error from "console" below):

naliboff commented:
@jdannberg - thanks for the fix. Ready to merge after resolving the merge conflict with the change log.

jdannberg commented:
Okay! Fixed the changelog as well.
